File: src/data/courses/dsa/queues/queue-stack-conversion/queue-using-stacks/code/python/queue_using_stacks.py
class QueueUsingStacks:
    def __init__(self):
        self.s1 = []
        self.s2 = []
    
    def enqueue(self, x):
        self.s1.append(x)
    
    def dequeue(self):
        if not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2.pop()
    
    def peek(self):
        if not self.s2:
            while self.s1:
                self.s2.append(self.s1.pop())
        return self.s2[-1]

File: src/data/courses/dsa/queues/queue-stack-conversion/queue-using-stacks/code/cpp/QueueUsingStacks.cpp
#include <stack>
using namespace std;
// ============================================================
// Solution 1: Two stacks — O(N) dequeue (transfer each time)
// ============================================================
class Solution1 {
    stack<int> in, out;
public:
    void enqueue(int x) { in.push(x); }
    int dequeue() {
        // Transfer ALL elements every dequeue — O(N)
        while(!in.empty()){out.push(in.top());in.pop();}
        int v=out.top();out.pop();
        while(!out.empty()){in.push(out.top());out.pop();}
        return v;
    }
    int front() {
        while(!in.empty()){out.push(in.top());in.pop();}
        int v=out.top();
        while(!out.empty()){in.push(out.top());out.pop();}
        return v;
    }
};

// ============================================================
// Solution 2: Two stacks — amortized O(1) dequeue (lazy transfer)
// ============================================================
class Solution2 {
    stack<int> in, out;
public:
    void enqueue(int x) { in.push(x); }
    int dequeue() {
        if(out.empty()) while(!in.empty()){out.push(in.top());in.pop();} // Transfer only when needed
        int v=out.top();out.pop();return v;
    }
    int front() {
        if(out.empty()) while(!in.empty()){out.push(in.top());in.pop();}
        return out.top();
    }
    bool empty() { return in.empty()&&out.empty(); }
};
